graph/structure/RuleStructure.java

```diff
@@ -51,5 +52,13 @@
      * Commits this {@code RuleStructure} to be persisted onto storage.
      */
     void commit();
+
+    void createConcludingIsaIndex(Label label);
+
+    void clearConcludingIsaIndex(Label label);
+
+    void createConcludingHasAttributeIndex(Label type);
+
+    void clearConcludesHasAttributeIndex(Label type);
```

Comment:
This is where we "inject" higher-level inferences down into the graph, via the rule structure.